diesel/src/sqlite/connection/raw.rs

```diff
@@ -292,99 +343,155 @@ extern "C" fn run_aggregator_step_function<ArgsSqlType, RetSqlType, Args, Ret, A
-        );
-        let mut aggregate_context = NonNull::new(aggregate_context as *mut OptionalAggregator<A>);
-        let aggregator = match aggregate_context.map(|a| &mut *a.as_ptr()) {
+        ffi::sqlite3_aggregate_context(ctx, std::mem::size_of::<OptionalAggregator<A>>() as i32)
+    };
+    let mut aggregate_context = NonNull::new(aggregate_context as *mut OptionalAggregator<A>);
+    let aggregator = unsafe {
+        match aggregate_context.map(|a| &mut *a.as_ptr()) {
             Some(&mut OptionalAggregator::Some(ref mut agg)) => agg,
             Some(mut a_ptr @ &mut OptionalAggregator::None) => {
                 ptr::write_unaligned(a_ptr as *mut _, OptionalAggregator::Some(A::default()));
                 if let &mut OptionalAggregator::Some(ref mut agg) = a_ptr {
                     agg
                 } else {
-                    unreachable!(
-                        "We've written the aggregator above to that location, it must be there"
-                    )
+                    assert_fail!("We've written the aggregator to the aggregate context, but it could not be retrieved.");
                 }
             }
             None => {
                 null_aggregate_context_error(ctx);
                 return;
             }
-        };
-
-        let mut f = |args: &[*mut ffi::sqlite3_value]| -> Result<(), Error> {
-            let args = build_sql_function_args::<ArgsSqlType, Args>(args)?;
-
-            Ok(aggregator.step(args))
-        };
+        }
+    };
 
-        let args = slice::from_raw_parts(value_ptr, num_args as _);
-        match f(args) {
-            Err(e) => {
-                let msg = e.to_string();
-                ffi::sqlite3_result_error(ctx, msg.as_ptr() as *const _, msg.len() as _);
-            }
-            _ => (),
-        };
-    }
+    let args = unsafe { slice::from_raw_parts(value_ptr, num_args as _) };
+    let args = build_sql_function_args::<ArgsSqlType, Args>(args);
+    let mut aggregator = std::panic::AssertUnwindSafe(aggregator);
+    let result = args
+        .map(|args| std::panic::catch_unwind(move || Ok(aggregator.step(args))))
+        .unwrap_or_else(|e| Ok(Err(e)));
+    match result {
+        Ok(Ok(())) => (),
+        Ok(Err(e)) => {
+            let msg = e.to_string();
+            unsafe { context_error_str(ctx, &msg) };
+        }
+        Err(_) => {
+            let msg = format!("{}::step() panicked", std::any::type_name::<A>());
+            unsafe { context_error_str(ctx, &msg) };
+        }
+    };
 }
 
 extern "C" fn run_aggregator_final_function<ArgsSqlType, RetSqlType, Args, Ret, A>(
     ctx: *mut ffi::sqlite3_context,
 ) where
-    A: SqliteAggregateFunction<Args, Output = Ret> + 'static + Send,
-    Args: FromSqlRow<ArgsSqlType, Sqlite>,
+    A: SqliteAggregateFunction<Args, Output = Ret> + 'static + Send + std::panic::UnwindSafe,
+    Args: FromSqlRow<ArgsSqlType, Sqlite> + std::panic::UnwindSafe,
```

Comment:
We do not even construct the arguments here, so `UnwindSafe` is not required here.